src/core_codemods/use_walrus_if.py

```diff
@@ -27,52 +34,85 @@ class UseWalrusIf(SemgrepCodemod):
         }
     ]
 
-    @classmethod
-    def rule(cls):
-        return """
-        rules:
-          - patterns:
-            - pattern: |
-                $ASSIGN
-                if $COND:
-                  $BODY
-            - metavariable-pattern:
-                metavariable: $ASSIGN
-                patterns:
-                  - pattern: $VAR = $RHS
-                  - metavariable-pattern:
-                      metavariable: $COND
-                      patterns:
-                        - pattern: $VAR
-                  - metavariable-pattern:
-                      metavariable: $BODY
-                      pattern: $VAR
-            - focus-metavariable: $ASSIGN
-        """
-
-    _modify_next_if: List[Tuple[CodeRange, cst.Assign]]
-    _if_stack: List[Optional[Tuple[CodeRange, cst.Assign]]]
+    _modify_next_if: List[Tuple[CodeRange, cst.NamedExpr]]
+    _if_stack: List[Optional[Tuple[CodeRange, cst.NamedExpr]]]
+    assigns: dict[cst.Assign, cst.NamedExpr]
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._modify_next_if = []
         self._if_stack = []
+        self.assigns = {}
+
+    def _build_named_expr(self, target, value, parens=True):
+        return cst.NamedExpr(
+            target=target,
+            value=value,
+            lpar=[cst.LeftParen()] if parens else [],
+            rpar=[cst.RightParen()] if parens else [],
+        )
 
-    def visit_If(self, node):
+    def on_visit(self, node: cst.CSTNode) -> Optional[bool]:
+        if len(node.children) < 2:
+            return super().on_visit(node)
+
+        for a, b in pairwise(node.children):
+            match (a, b):
+                case (
+                    cst.SimpleStatementLine(
+                        body=[
+                            cst.Assign(
+                                targets=[
+                                    cst.AssignTarget(target=cst.Name() as target),
+                                ],
+                                value=value,
+                            ) as assign
+                        ]
+                    ),
+                    cst.If(test=test),
+                ):
+                    # If test can be a comparison expression
+                    match test:
+                        case cst.Comparison(
+                            left=cst.Name() as left,
+                            comparisons=[
+                                cst.ComparisonTarget(
+                                    operator=(
+                                        cst.Is()
+                                        | cst.IsNot()
+                                        | cst.Equal()
+                                        | cst.NotEqual()
+                                    )
+                                )
+                            ],
+                        ):
+                            # TODO: use actual scope analysis?
```

Comment:
Since you guarantee those are consecutive statements, target.value there is no need to do a scope analysis.

Reply:
👍 that's what I was thinking too. There's definitely a more sophisticated version of this codemod that identifies non-adjacent statements using scope analysis. But that's obviously more difficult to implement and I also think it's a lot less safe in general since reordering statements can have unanticipated side effects. I'm happy with this restriction for now.